dbms/src/Flash/Mpp/ReceivedMessageQueue.cpp

```diff
@@ -0,0 +1,150 @@
+                /// if there is no remaining consumer, then pop it from original queue, the message must stay in the queue before the pop
+                /// so even use tryPop, the result must not be empty
+                assert(pop_result != MPMCQueueResult::EMPTY);
+                if (original_msg != nullptr)
+                    RUNTIME_CHECK_MSG(*original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
+            }
+        }
+    }
+    else
+    {
+        if constexpr (need_wait)
+        {
+            res = grpc_recv_queue->pop(recv_msg);
+        }
+        else
+        {
+            res = grpc_recv_queue->tryPop(recv_msg);
+        }
+    }
+    return {res, recv_msg};
+}
+
+template <bool is_force>
+bool ReceivedMessageQueue::pushToMessageChannel(ReceivedMessagePtr & received_message, ReceiverMode mode)
+{
+    bool success = false;
+    if constexpr (is_force)
+        success = msg_channel->forcePush(received_message) == MPMCQueueResult::OK;
+    else
+        success = msg_channel->push(received_message) == MPMCQueueResult::OK;
+
+    injectFailPointReceiverPushFail(success, mode);
+    return success;
+}
+
+GRPCReceiveQueueRes ReceivedMessageQueue::pushToGRPCReceiveQueue(ReceivedMessagePtr & received_message)
+{
+    auto res = grpc_recv_queue->push(received_message);
+    fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, res = GRPCReceiveQueueRes::CANCELLED);
+    return res;
+}
+
+ReceivedMessageQueue::ReceivedMessageQueue(
+    const AsyncRequestHandlerWaitQueuePtr & conn_wait_queue,
+    const LoggerPtr & log_,
+    size_t max_buffer_size,
+    bool enable_fine_grained,
+    size_t fine_grained_channel_size_)
+    : fine_grained_channel_size(enable_fine_grained ? fine_grained_channel_size_ : 0)
+    , log(log_)
+{
+    if (enable_fine_grained)
+    {
+        for (size_t i = 0; i < fine_grained_channel_size; ++i)
+            /// these are unbounded queues
+            msg_channels_for_fine_grained_shuffle.push_back(std::make_shared<LooseBoundedMPMCQueue<ReceivedMessagePtr>>(std::numeric_limits<size_t>::max()));
+        /// use pushcallback to make sure that the order of messages in msg_channels_for_fine_grained_shuffle is exactly the same as it in msg_channel,
+        /// because pop from msg_channel rely on this assumption. An alternative is to make msg_channel a set/map of messages for fine grained shuffle, but
+        /// it need many more changes
+        msg_channel = std::make_shared<LooseBoundedMPMCQueue<ReceivedMessagePtr>>(max_buffer_size, [this](const ReceivedMessagePtr & element) {
+            for (size_t i = 0; i < fine_grained_channel_size; ++i)
+            {
+                if (msg_channels_for_fine_grained_shuffle[i]->tryPush(element) != MPMCQueueResult::OK)
+                    return false;
+            }
+            return true;
+        });
```

Comment:
why not moving the `push_callback` to `pushToMessageChannel`?
Adding a `push_callback` to `LooseBoundMPMCQueue` is a bit weird

Reply:
Moving `push_callback` to `pushToMessageChannel` is not enough under currently implemantion. The main idea of current implementation is to use `msg_channel` as a bounded queue, it only handle block `push`, the consumer streams does not `pop` messages from `msg_channel` directly, instead, they `pop` messages from `fine_grained_msg_channels`,  and only if a message is poped by all the consumers(remaining_consumers == 0), it will be poped from `msg_channel`. Since the queue is a FIFO queue, there is actually no way to pop a specific message from a queue, the hack thing is we keep the order of messages the same in all the queues, then we can ensure that if a message has no remaining consumers, it must be the earliest message in `msg_channel`, so it can be safely poped from `msg_channel`. If the order of all the messages is not guaranteed, `msg_channel` need to be a message set instead of a message queue, and every message need a unique key, once a message is consumed by all the consumers, it can be moved from the message set by its unique key.